### ISSUE TITLE
We should add PTC2

### ISSUE THREAD

**Issue by dvpfagan** [jmmcd/PonyGE2#63]
We should add PTC2-[See here](http://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=5586336)
Robin Harper applied it to GE and Sean Luke had previously used it in GP. 

If we are striving to make PonyGE state of the art we should look to include stuff like this

Dave

**Comment by mfjoneill** [comment]
We should definitely add PTC2, and also to a linear random initialisation add an enhanced variant which removes duplicates and invalids. Miguel's recent paper shows the superiority of these two initialisers (using linear operators) over all other variants
https://link.springer.com/article/10.1007/s10710-017-9309-9

Many thanks,
Mike.

**Comment by jmmcd** [comment]
Added Nicolau's RVD.

PTC2 would be more effort, will not attempt for now.

**Comment by jmmcd** [comment]
According to Nicolau: "Harper [8] used PTC2 to initialise individuals in GE, with a uniform distribution of (derivation) tree-sizes; in his implementation, the size of a derivation tree is the number of grammar expansions performed."

Maybe we could get most of the benefit of PTC2 but still without looking at trees, as follows. We measure the *size* (not *depth*) of a tree in the same way as Harper. We use a discard approach (same as RVD) to make sure that we achieve the desired distribution of tree sizes.

**Comment by jmmcd** [comment]
![init_results](https://user-images.githubusercontent.com/167942/137646977-3b1904a7-d2e1-4bf1-9539-a38c0bf3a753.png)

**Comment by dvpfagan** [comment]
Makes sense to try get PTC2 or a good approximation in the codebase. From
memory, Robins ultimate goal was towards evolving/coevolving code/behaviour
or robots. I think the other approaches were mainly tested on
regression-based problems (I am also aware some people regard all problems
as a form of regression problems) Anyway it think ptc2 or some equivalent
would be great.
